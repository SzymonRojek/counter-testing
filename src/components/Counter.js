import { useEffect, useState } from "react";

import "./styles.css";
export const Counter = ({ start }) => {
  const [defaultStart, setDefaultStart] = useState(start);
  const [counter, setCounter] = useState(0);
  const [inputValue, setInputValue] = useState("");

  useEffect(() => {
    setDefaultStart(start);
  }, [start]);

  const additionHandler = () =>
    start
      ? setDefaultStart((prev) => prev + 1)
      : setCounter((prev) => prev + 1);

  const subtractionHandler = () =>
    start
      ? setDefaultStart((prev) => prev - 1)
      : setCounter((prev) => prev - 1);

  const handleChangeInputValue = ({ target }) => setInputValue(target.value);

  const handleReset = () => {
    setDefaultStart(start);
    setCounter(0);
    setInputValue("");
  };

  const handleSubmit = (e) => {
    e.preventDefault();

    if (/\d/.test(inputValue)) {
      setCounter(+inputValue);
      setDefaultStart(+inputValue);
    }

    setInputValue("");
  };

  return (
    <div>
      <p>
        Counter result:
        <span className="counter">{`${start ? defaultStart : counter}`}</span>
      </p>

      <button className="addition-button one" onClick={additionHandler}>
        +
      </button>
      <button className="subtraction-button two" onClick={subtractionHandler}>
        -
      </button>

      <form onSubmit={handleSubmit}>
        <label>
          <input
            className="number-input"
            placeholder="change counter"
            type="number"
            value={inputValue}
            onChange={handleChangeInputValue}
          ></input>
        </label>
        <button type="submit" className="change-button three">
          Change
        </button>
        <button
          className="reset-button four"
          type="reset"
          onClick={handleReset}
        >
          Reset
        </button>
        {""}
      </form>
    </div>
  );
};
