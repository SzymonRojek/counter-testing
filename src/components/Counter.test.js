import renderer from "react-test-renderer";
import { shallow } from "enzyme";

import { Counter } from "./Counter";

it("renders correctly", () => {
  const tree = renderer.create(<Counter />).toJSON();
  expect(tree).toMatchSnapshot();
});

describe("Counter component", () => {
  it("should renders all buttons elements", () => {
    const wrapper = shallow(<Counter />);
    expect(wrapper.find('[data-test="text"]').html()).toContain(
      "Counter result:"
    );

    const additionButton = wrapper.find('[data-test="add-btn"]');
    expect(additionButton.text()).toBe("+");

    const subtractionButton = wrapper.find('[data-test="sub-btn"]');
    expect(subtractionButton.text()).toBe("-");

    const changeButton = wrapper.find('[data-test="change-btn"]');
    expect(changeButton.text()).toBe("Change");

    const resetButton = wrapper.find('[data-test="reset-btn"]');
    expect(resetButton.text()).toBe("Reset");
  });

  it("should state be 0 when props start is not added", () => {
    const wrapper = shallow(<Counter />);
    const counter = wrapper.find('[data-test="counter"]');
    expect(counter.text()).toBe("0");
  });

  it("should state be 10 when props start is added", () => {
    const wrapper = shallow(<Counter start={10} />);
    const counter = wrapper.find('[data-test="counter"]');
    expect(counter.text()).toBe("10");
  });

  it("should increment counter from 0 to 1 when default counter is 0", () => {
    const wrapper = shallow(<Counter />);
    const additionButton = wrapper.find('[data-test="add-btn"]');
    let counter = wrapper.find('[data-test="counter"]');
    expect(counter.text()).toBe("0");

    additionButton.simulate("click");
    counter = wrapper.find('[data-test="counter"]');
    expect(counter.text()).toBe("1");

    expect(wrapper.find('[data-test="text"]').html()).toContain(
      "Counter result:"
    );
    expect(wrapper.find('[data-test="counter"]').html()).toContain("1");
  });

  it("should decrement counter from 0 to -1 when default counter is 0", () => {
    const wrapper = shallow(<Counter />);
    const subtractionButton = wrapper.find('[data-test="sub-btn"]');
    let counter = wrapper.find('[data-test="counter"]');
    expect(counter.text()).toBe("0");

    subtractionButton.simulate("click");
    counter = wrapper.find('[data-test="counter"]');
    expect(counter.text()).toBe("-1");
  });

  it("should increment counter from 10 to 13 when props start is 10", () => {
    const wrapper = shallow(<Counter start={10} />);
    const additionButton = wrapper.find('[data-test="add-btn"]');
    let counter = wrapper.find('[data-test="counter"]');
    expect(counter.text()).toBe("10");

    additionButton.simulate("click");
    additionButton.simulate("click");
    additionButton.simulate("click");
    counter = wrapper.find('[data-test="counter"]');
    expect(counter.text()).toBe("13");
  });

  it("should decrement counter from 10 to 7 when props start is 10", () => {
    const wrapper = shallow(<Counter start={10} />);
    const subtractionButton = wrapper.find('[data-test="sub-btn"]');
    let counter = wrapper.find('[data-test="counter"]');
    expect(counter.text()).toBe("10");

    subtractionButton.simulate("click");
    subtractionButton.simulate("click");
    subtractionButton.simulate("click");
    counter = wrapper.find('[data-test="counter"]');
    expect(counter.text()).toBe("7");
  });

  it("should input change a value when number is typed, for example 27", () => {
    const wrapper = shallow(<Counter />);
    const numberInput = wrapper.find('[data-test="input-num"]');
    numberInput.simulate("change", {
      target: { value: "27" },
    });

    expect(wrapper.find("input").get(0).props.value).toEqual("27");
  });

  it("should counter change on submit when the change button is clicked - no props start added", () => {
    const wrapper = shallow(<Counter />);
    let counter = wrapper.find('[data-test="counter"]');
    expect(counter.text()).toBe("0");

    const numberInput = wrapper.find('[data-test="input-num"]');
    numberInput.simulate("change", {
      target: { value: "27" },
    });

    counter = wrapper.find('[data-test="counter"]');
    expect(counter.text()).toBe("0");

    wrapper.find("form").simulate("submit", {
      preventDefault: () => {},
    });
    const submitButton = wrapper.find('button[type="submit"]');
    submitButton.simulate("click");
    counter = wrapper.find('[data-test="counter"]');
    expect(counter.text()).toBe("27");
  });

  it("should counter change on submit when change button is clicked - props start added", () => {
    const wrapper = shallow(<Counter start={10} />);
    let counter = wrapper.find('[data-test="counter"]');
    expect(counter.text()).toBe("10");

    const numberInput = wrapper.find('[data-test="input-num"]');
    numberInput.simulate("change", {
      target: { value: "50" },
    });

    counter = wrapper.find('[data-test="counter"]');
    expect(counter.text()).toBe("10");

    wrapper.find("form").simulate("submit", {
      preventDefault: () => {},
    });

    const submitButton = wrapper.find('button[type="submit"]');
    submitButton.simulate("click");

    counter = wrapper.find('[data-test="counter"]');
    expect(counter.text()).toBe("50");
  });

  it("should get initial value 0 when the reset button is clicked - no props start added", () => {
    const wrapper = shallow(<Counter />);
    let counter = wrapper.find('[data-test="counter"]');
    expect(counter.text()).toBe("0");

    const numberInput = wrapper.find('[data-test="input-num"]');
    numberInput.simulate("change", {
      target: { value: "1227" },
    });

    counter = wrapper.find('[data-test="counter"]');
    expect(counter.text()).toBe("0");

    wrapper.find("form").simulate("submit", {
      preventDefault: () => {},
    });

    const submitButton = wrapper.find('button[type="submit"]');
    submitButton.simulate("click");

    counter = wrapper.find('[data-test="counter"]');
    expect(counter.text()).toBe("1227");

    const resetButton = wrapper.find('button[type="reset"]');
    resetButton.simulate("click");

    counter = wrapper.find('[data-test="counter"]');
    expect(counter.text()).toBe("0");
  });

  it("should get initial value 10 when the reset button is clicked - props start added", () => {
    const wrapper = shallow(<Counter start={10} />);
    let counter = wrapper.find('[data-test="counter"]');
    expect(counter.text()).toBe("10");

    const numberInput = wrapper.find('[data-test="input-num"]');
    numberInput.simulate("change", {
      target: { value: "300" },
    });

    counter = wrapper.find('[data-test="counter"]');
    expect(counter.text()).toBe("10");

    wrapper.find("form").simulate("submit", {
      preventDefault: () => {},
    });

    const submitButton = wrapper.find('button[type="submit"]');
    submitButton.simulate("click");

    counter = wrapper.find('[data-test="counter"]');
    expect(counter.text()).toBe("300");

    const resetButton = wrapper.find('button[type="reset"]');
    resetButton.simulate("click");

    counter = wrapper.find('[data-test="counter"]');
    expect(counter.text()).toBe("10");
  });
});
