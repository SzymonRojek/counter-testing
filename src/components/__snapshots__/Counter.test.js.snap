// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <p>
    Counter result:
    <span
      className="counter"
    >
      0
    </span>
  </p>
  <button
    className="addition-button one"
    onClick={[Function]}
  >
    +
  </button>
  <button
    className="subtraction-button two"
    onClick={[Function]}
  >
    -
  </button>
  <form
    onSubmit={[Function]}
  >
    <label>
      <input
        className="number-input"
        onChange={[Function]}
        placeholder="change counter"
        type="number"
        value=""
      />
    </label>
    <button
      className="change-button three"
      type="submit"
    >
      Change
    </button>
    <button
      className="reset-button four"
      onClick={[Function]}
      type="reset"
    >
      Reset
    </button>
    
  </form>
</div>
`;
